### ISSUE TITLE
Is it safe to reconfigure <zookeeper> on running deployment?

### ISSUE THREAD

**Issue by yukha-dw** [ClickHouse/ClickHouse#68318]
### Question

Hello, I'm running CH Server + standalone CH Keeper 24.3.6.48 LTS

I want to replace one node of CH Keeper cluster from:
```
server.1=chk-100:9444;participant;1
server.2=chk-200:9444;participant;1
server.3=chk-300:9444;participant;1
```
to this:
```
server.1=chk-100:9444;participant;1
server.2=chk-200:9444;participant;1
server.4=chk-400:9444;participant;1
```

I found no error message when reconfiguring `<keeper_server>` on CH Keeper. But as soon as I reconfigure `<zookeeper>` on CH Server, my CH nodes lose connection to ZooKeeper for a moment.

```JSON
{"date_time":"1723612261.418051","thread_name":"","thread_id":"360","level":"Error","query_id":"","logger_name":"sampelDB.sampleTable (ffffffff-ffff-ffff-ffff-ffffffffffff)","message":"void DB::StorageReplicatedMergeTree::queueUpdatingTask(): Code: 999. Coordination::Exception: Session expired. (KEEPER_EXCEPTION), Stack trace (when copying this message, always include the lines below):\n\n0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000cbdb9fb\n1. Coordination::Exception::Exception<char const (&) [16]>(char const (&) [16], Coordination::Error) @ 0x0000000012adf4bb\n2. Coordination::ZooKeeper::pushRequest(Coordination::ZooKeeper::RequestInfo&&) @ 0x0000000012b08978\n3. Coordination::ZooKeeper::get(String const&, std::function<void (Coordination::GetResponse const&)>, std::shared_ptr<std::function<void (Coordination::WatchResponse const&)>>) @ 0x0000000012b0ac25\n4. zkutil::ZooKeeper::getImpl(String const&, String&, Coordination::Stat*, std::function<void (Coordination::WatchResponse const&)>) @ 0x0000000012ab913d\n5. zkutil::ZooKeeper::tryGetWatch(String const&, String&, Coordination::Stat*, std::function<void (Coordination::WatchResponse const&)>, Coordination::Error*) @ 0x0000000012ab985f\n6. zkutil::ZooKeeper::tryGet(String const&, String&, Coordination::Stat*, std::shared_ptr<Poco::Event> const&, Coordination::Error*) @ 0x0000000012ab9748\n7. zkutil::ZooKeeper::get(String const&, Coordination::Stat*, std::shared_ptr<Poco::Event> const&) @ 0x0000000012ab9598\n8. DB::ReplicatedMergeTreeQueue::pullLogsToQueue(std::shared_ptr<zkutil::ZooKeeper>, std::function<void (Coordination::WatchResponse const&)>, DB::ReplicatedMergeTreeQueue::PullLogsReason) @ 0x000000001208d599\n9. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<DB::StorageReplicatedMergeTree::StorageReplicatedMergeTree(String const&, String const&, DB::LoadingStrictnessLevel, DB::StorageID const&, String const&, DB::StorageInMemoryMetadata const&, std::shared_ptr<DB::Context>, String const&, DB::MergeTreeData::MergingParams const&, std::unique_ptr<DB::MergeTreeSettings, std::default_delete<DB::MergeTreeSettings>>, DB::RenamingRestrictions, bool)::$_1, void ()>>(std::__function::__policy_storage const*) @ 0x00000000119a0ba4\n10. DB::BackgroundSchedulePool::threadFunction() @ 0x000000000fdaff07\n11. void std::__function::__policy_invoker<void ()>::__call_impl<std::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<false>::ThreadFromGlobalPoolImpl<DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, char const*)::$_0>(DB::BackgroundSchedulePool::BackgroundSchedulePool(unsigned long, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, StrongTypedef<unsigned long, CurrentMetrics::MetricTag>, char const*)::$_0&&)::'lambda'(), void ()>>(std::__function::__policy_storage const*) @ 0x000000000fdb1073\n12. void* std::__thread_proxy[abi:v15000]<std::tuple<std::unique_ptr<std::__thread_struct, std::default_delete<std::__thread_struct>>, void ThreadPoolImpl<std::thread>::scheduleImpl<void>(std::function<void ()>, Priority, std::optional<unsigned long>, bool)::'lambda0'()>>(void*) @ 0x000000000cc8a74d\n13. ? @ 0x00007fddc0a6aac3\n14. ? @ 0x00007fddc0afc850\n (version 24.3.6.48 (official build))","source_file":"src/Common/Exception.cpp; void DB::tryLogCurrentExceptionImpl(Poco::Logger *, const std::string &)","source_line":"222"}
```

My concern are:
1. Do I have to manually restart CH Server / Keeper to finalize the configuration?
2. During `<zookeeper>` reconfiguration, does CH Server work normally without any downtime?

Thank you!

**Comment by den-crane** [comment]
>CH Server, my CH nodes lose connection to ZooKeeper for a moment.

And what ? I don't see a problem. CH will reestablish connection.

>Do I have to manually restart CH Server / Keeper to finalize the configuration?

Yes, you have to.

>During <zookeeper> reconfiguration, does CH Server work normally without any downtime?

Yes, it does.

**Comment by yukha-dw** [comment]
Thanks for your response @den-crane!

> And what ? I don't see a problem. CH will reestablish a connection.

I saw this related [guide](https://kb.altinity.com/altinity-kb-setup-and-maintenance/altinity-kb-server-config-files/#server-config-configxml-sections-which-dont-require-restart) written by Altinity. It said reconnection doesn't happen automatically, so I guess it's hasn't been updated to current LTS behavior then

> Do I have to manually restart CH Server / Keeper to finalize the configuration?
> > Yes, you have to.

I've tested on local without restart and results are vary, sometimes CH Keeper uses new configuration and sometimes CH Keeper keeps trying to reconnect with replaced node.
